tests/gen/gen_auto.go

```diff
@@ -0,0 +1,245 @@
+	configurator docsGenConfigurator
+
+	generatedDocKeys map[string][]string
+	random           rand.Rand
+}
+
+func (g *randomDocGenerator) generateDocs(options ...Option) ([]GeneratedDoc, error) {
+	err := g.configurator.Configure(options...)
+	if err != nil {
+		return nil, err
+	}
+
+	g.random = *g.configurator.random
+
+	resultDocs := make([]GeneratedDoc, 0, g.getMaxTotalDemand())
+	docsLists, err := g.generateRandomDocs(g.configurator.typesOrder)
+	if err != nil {
+		return nil, err
+	}
+	for _, docsList := range docsLists {
+		typeDef := g.configurator.types[docsList.ColName]
+		for _, doc := range docsList.Docs {
+			resultDocs = append(resultDocs, GeneratedDoc{
+				ColName: typeDef.Description.Name,
+				JSON:    createDocJSON(&typeDef, doc),
+			})
+		}
+	}
+	return resultDocs, nil
+}
+
+func (g *randomDocGenerator) getMaxTotalDemand() int {
+	totalDemand := 0
+	for _, demand := range g.configurator.docsDemand {
+		totalDemand += demand.max
+	}
+	return totalDemand
+}
+
+// getNextPrimaryDocKey returns the key of the next primary document to be used as a relation.
+func (g *randomDocGenerator) getNextPrimaryDocKey(secondaryType string, field *client.FieldDescription) string {
+	ind := g.configurator.usageCounter.getNextTypeIndForField(secondaryType, field)
+	docKey := g.generatedDocKeys[field.Schema][ind]
+	return docKey
+}
+
+func (g *randomDocGenerator) getDocKey(typeDef *client.CollectionDefinition, doc map[string]any) (string, error) {
+	clientDoc, err := client.NewDocFromJSON([]byte(createDocJSON(typeDef, doc)))
+	if err != nil {
+		return "", err
+	}
+	return clientDoc.Key().String(), nil
+}
+
+func (g *randomDocGenerator) generateRandomDocs(order []string) ([]DocsList, error) {
+	result := []DocsList{}
+	for _, typeName := range order {
+		col := DocsList{ColName: typeName}
+		typeDef := g.configurator.types[typeName]
+
+		currentTypeDemand := g.configurator.docsDemand[typeName]
+		averageDemand := currentTypeDemand.getAverage()
+		for i := 0; i < averageDemand; i++ {
+			newDoc := make(map[string]any)
+			for _, field := range typeDef.Schema.Fields {
+				if field.Name == request.KeyFieldName {
+					continue
+				}
+				if field.IsRelation() {
+					if field.IsPrimaryRelation() {
+						newDoc[field.Name+request.RelatedObjectID] = g.getNextPrimaryDocKey(typeName, &field)
+					}
+				} else {
+					fieldConf := g.configurator.config.ForField(typeName, field.Name)
+					newDoc[field.Name] = g.generateRandomValue(typeName, field.Kind, fieldConf)
+				}
+			}
+			docKey, err := g.getDocKey(&typeDef, newDoc)
+			if err != nil {
+				return nil, err
+			}
+			g.generatedDocKeys[typeName] = append(g.generatedDocKeys[typeName], docKey)
+			col.Docs = append(col.Docs, newDoc)
+		}
+		result = append(result, col)
+	}
+	return result, nil
+}
+
+func getRandomString(random *rand.Rand, n int) string {
+	const letterBytes = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+	b := make([]byte, n)
+	for i := range b {
```

Comment:
nitpick: I think random number generation is relatively slow, and apart from the risk of collsion, there does not seem to be a good reason to independently randomise each char in the string, you could just generate a random start index, and optionally a randomized offset, then construct the string from that:

```
charIndex := random.Intn(len(letterBytes))
offset := random.Intn(len(letterBytes)) // or whatever :) 
for i := range b {
   b[i] = letterBytes[charIndex]
   charIndex += offset // You'll need to wrap this, I'm too lazy to write that here
}
```
Marked as a nitpick, as I'm not sure how significant the performance difference would be given normal sized documents and the relative cost of inserting them into Defra etc.

Perhaps an even better approach would to only randomize the initial string in the  `generateDocs` call, and then just increment the characters for subsequent strings - then there would be no risk of collision, and only one random call in the batch.

Reply:
feels this this comment slightly contradicts the other comment you left regarding collisions https://github.com/sourcenetwork/defradb/pull/2051/files/26ef63c132318f71007f4cb0a1710e3354043f52#r1401023373

With the approach that you proposed, yes, we get getter performance but at the expense of randomness, which will drastically increase chances for collision.

Lets take a string of length 8. With the current approach there are 52^8 permutations available. With your approach it drop to 52^2 which is just 2704.

Moreover, I don't think it should be hard-coded into the framework to set a level of randomness. It should be up to the client to decide. If the client is concerned about the performance he can just require string length to be a certain number (with number 2 it will call random number generator twice). If the client wants to have a longer name, he can define a decorator on top of the generated string:
```go
WithFieldGenerator("User", "name", func(i int, next func() any) any {
	return "some_prefix_" + next().(string)
}),
```
The performance concern arises mostly if a large set of document needs to be generated, but in this case you surely need a uniformly randomized string.

Reply:
> The performance concern arises mostly if a large set of document needs to be generated, but in this case you surely need a uniformly randomized string.

Why?

More collisions can still be cheaper than a complex randomization algo.

And, are you against the below suggestion, and if so why?:

> Perhaps an even better approach would to only randomize the initial string in the generateDocs call, and then just increment the characters for subsequent strings - then there would be no risk of collision, and only one random call in the batch.

Reply:
While more collisions might be cheaper computationally, they can compromise the integrity of the system if uniqueness is a requirement. The cost of handling collisions (both in terms of performance and complexity) might outweigh the benefits of a simpler randomization algorithm.

> Perhaps an even better approach would to only randomize the initial string in the generateDocs call, and then just increment the characters for subsequent strings - then there would be no risk of collision, and only one random call in the batch.

I've actually thought about this option myself but more as an additional option for clients to chose between randomness vs performance.